docs/configuration.md
# Configuration
Serverless Sharp stores its settings in YAML files. When deploying the application, you specify the YAML file and the
stage (your environment). For example:

```shell-session
serverless deploy --settings mysite.com.yml --stage prod
```

## Settings File Structure
We use YML files to make use of their inheritance. This allows us to define some defaults and then branch environment
-specific changes off of them. You can define multiple environments, each of which will inherit your default settings. 
This is useful if you have different buckets for staging & production. You may also wish to create an
 environment without a sign-key for local development. For example:

```yaml
defaults: &defaults
  serviceName: image-handler
  region: 'us-east-1'
  environment: &defaults.environment
    DEFAULT_QUALITY: 75
    DEFAULT_COMPRESS_QUALITY: 45
    SLS_IGNORE: favicon.ico
    SLS_VALID_PATH_REGEX: '.*'
    MAX_IMAGE_WIDTH: 2000
    MAX_IMAGE_HEIGHT: 1000
    PNGQUANT_SPEED: 10
    SOURCE_BUCKET: 'images.mydomain.com'
    DEFAULT_CACHE_CONTROL: 'max-age=2592000'

stages:
  dev:
    <<: *defaults
    environment:
      <<: *defaults.environment
      SECURITY_KEY: ''
      CUSTOM_DOMAIN: ''
      ACM_CERTIFICATE_ARN: ''
  prod:
    <<: *defaults
    environment:
      <<: *defaults.environment
      SECURITY_KEY: 'random-string-here'
      CUSTOM_DOMAIN: 'img.mydomain.com'
      ACM_CERTIFICATE_ARN: ''
```

## Environment Settings
The majority of the application and environment specific settings are configured as environment variables that are
 passed into the Lambda service handler.
 
 |Setting|Default|
 |-|-|
 |[CUSTOM_DOMAIN](#custom_domain)|*empty*|
 |[ACM_CERTIFICATE_ARN])(#acm_certificate_arn)|*empty*]
 |[DEFAULT_CACHE_CONTROL](#default_cache_control)|*empty*|
 |[DEFAULT_COMPRESS_QUALITY](#default_compress_quality)|75|
 |[DEFAULT_QUALITY](#default_quality)|75|
 |[MAX_IMAGE_HEIGHT](#max_image_height)|`2000`|
 |[MAX_IMAGE_WIDTH](#max_image_width)|`2000`|
 |[PNGQUANT_SPEED](#pngquant_speed)|`10`|
 |[SECURITY_KEY](#security_key)|*empty*|
 |[SOURCE_BUCKET](#source_bucket)|*empty*|
 |[SLS_IGNORE](#sls_ignore)|*empty*|
 |[SLS_VALID_PATH_REGEX](#sls_valid_path_regex)|`.*`|
 

### `CUSTOM_DOMAIN`

- Type: `string`
- Default: *empty*

If set, Serverless Sharp will generate an Amazon issued certificate for the domain and link it to your Cloudfront
 distribution.

<Note type="tip">

This requires deployment in the `us-east-1` region!

</Note>

<Note type="tip">

During deployment, you'll need to go into your Amazon ACM console and verify the domain.

</Note>

### `ACM_CERTIFICATE_ARN`
- type: `string`
- DEFAULT: *empty*

Use an existing ACM Certificate, by supplying the ACM Certificate Arn.

### `DEFAULT_CACHE_CONTROL`

- Type: `string`
- Default: *empty*

By default, Serverless Sharp will serve responses with the same Cache-Control header as the object being served. If
 the Cache-Control header is not set, this value will be used.

### `DEFAULT_COMPRESS_QUALITY`

- Type: `integer`
- Default: `75`

A number between 1 and 100 for the default quality when an image is requested to be compressed

 ### `DEFAULT_QUALITY`
 
 - Type: `integer`
 - Default: `75`
 
 
### `MAX_IMAGE_HEIGHT`

- Type: `integer`
- Default: `2000`

Images may not be requested with a height larger than this value

### `MAX_IMAGE_WIDTH`

- Type: `integer`
- Default: `2000`

Images may not be requested with a width larger than this value

### `PNGQUANT_SPEED`

- Type: `integer`
- Default: `10`

The speed value to pass to the `pngquant` optimization program. From the [pngquant website](https://pngquant.org/):

> Speed/quality trade-off from 1 (brute-force) to 10 (fastest). The default is 3. Speed 10 has 5% lower quality, but is 8 times faster than the default.

### `SECURITY_KEY`

- Type: `string`
- Default: *empty*

If set, a hash must be provided as the `s` parameter in all requests. See [Security](/usage/security) for more.

### `SOURCE_BUCKET`

- Type: `string`
- Default: *empty*

The S3 bucket in your account where your images are stored - you can include a path here if you like. For example
: `mybucket/images`

This will tell Serverless to create an IAM role with access to all objects in the `mybucket` bucket with an object prefix
of `images`.

Serverless Sharp will then validate all requests to this prefix, if it doesn't start with that prefix, the system
 will automatically prepend it to the request. This means these requests are effectively equivalent:

`localhost/images/my-image.png`

`localhost/my-image.png`

### `SLS_IGNORE`

- Type: `string`
- Default: *empty*

A comma-delineated string of paths that should be ignored (for example, `favicon.ico`)

### `SLS_VALID_PATH_REGEX`

- Type: `Regular Expression`
- Default: `.*`

A regular expression for path validation. Allows you to explicitly control what assets may be handled by the Lambda
 function. If the request path fails this test, a 404 response will be served.

Example:
```yaml
# All requests must start with /images/
SLS_VALID_PATH_REGEX: ^\/images\/.*`
```


docs/usage/parameters.md
# Parameters

We chose to base our API around the [Imgix service](https://docs.imgix.com/apis/url) to allow for backwards compatibility
with the already popular service. The idea is that all CMS plugins should be able to seamlessly use this service in-place of
an Imgix URL. We've only implemented a hand-full of the features Imgix offers; however, the one's we've
implemented should cover most use-cases.

The benefits of using this method over other methods (such as hashing the entire URL payload in base64) are:
- Much more intuitive
- Easier to develop & debug
- Provides clear prefix matching your original object's path with which you can create invalidations with wildcards

You may access images in your `SOURCE_BUCKET` via the Cloudfront URL that is generated for your distribution just like
normal images. Transforms can be appended to the filename as query parameters.


## `fm` - output format 

Can be one of: `webp`, `png`, `jpeg`, `tiff`

## `w` - width

Scales image to supplied width while maintaining aspect ratio

## `h` - height 

Scales image to supplied height while maintaining aspect ratio

<Note type="tip">

*If both width and height are supplied, the aspect ratio will be preserved and scaled to minimum of either width/height*

</Note>

## `q` - quality 

(75) - 1-100

## `ar` - aspect-ratio 

(1.0:1.0) - When fit=crop, an aspect ratio such as 16:9 can be supplied, optionally with a
 height or width. If neither height or width are defined, the original image size will be used.
 
 
## `dpr` - device-pixel-ratio 

(1) - scales requested image dimensions by this multiplier.

## `fit` - resize fitting mode
 
Can be one of: `fill`, `scale`, `crop`, `clip`, `min`, `max`

## `fill-color`
 
used when `fit` is set to `fill` can be a loosely formatted color such as "red" or "rgb(255,0,0)"

## `crop` - resize fitting mode

can be one of: `focalpoint`, `entropy`, any comma separated combination of `top`, `bottom`, `left` `right`

### `crop=focalpoint`

Uses the `fp-x` and `fp-y` parameters to crop as close to the supplied point as possible.

### `crop=entropy`

Crops the image around the region with the highest [Shannon entropy](https://en.wikipedia.org/wiki/Entropy_%28information_theory%29) 

### `crop=top,left` (or `bottom`, `right`)

Crops the image around the region specified. Supply up to two region identifiers comma separated.

## `fp-x`, `fp-y` - focal point x & y
 
Percentage, 0 to 1 for where to focus on the image when cropping with focalpoint mode

## `s` - security hash 

See [Security](/usage/security.md#request-query-hashing) section

## `auto`

Can be a comma separated combination of: `compress`, `format`

### `auto=format`
If `auto` includes format, the service will try to determine the ideal format to convert the image to. The rules are:
- If the browser supports it, everything except for gifs is returned as webp
- If a png is requested and that png has no alpha channel, it will be returned as a jpeg

### `auto=compress`
The `compress` parameter will try to run post-processed optimizations on the image prior to returning it.
- `png` images will run through `pngquant`


## `blur` - gaussian blur

Between 0-2000


CHANGELOG.md
# Changelog

## Unreleased
### Fixed
- Fix quality parameter (`q`) not being respected with `auto=format`

## [2.1.1] - 2021-03-31
### Changed
- Updated to Sharp 0.28.0

## [2.1.0] - 2021-03-24

### Added
- Adds ACM_CERTIFICATE_ARN configuration variable (@tyrauber)

### Changed
- ImageRequest exceptions now throw S3Exception instead of generic Error (@marco-primiceri)
- No longer require serverless-apiw-binary and serverless-apigwy-binary (@tyrauber)
- Responses now include a Last-Modified header with the generation time for an image (@tyrauber)
- Security hash is no longer case-sensitive
- Index function now always returns a response to support middleware applications (@tyrauber)
- Lambda timeout increased from default (6 seconds) to 10 seconds
- Improve CI workflow (@dashmug)
- Update to Sharp 0.27.2
- Update Lambda runtime to Node 14
- Removed Serverless Offline plugin due to lack of Node 14 compatibility

### Fixed
- Fix potential issue processing font expectations (@tyrauber)
- Properly capture S3 exceptions (@marco-primiceri)
- pngquant was not working properly
- fit=max was not working properly (@kylecotter)
- Fix config for custom CloudFormation Outputs. (@dashmug)
- Don't process SVGs in format=auto (@kylecotter)

## [2.0.6] - 2020-05-13

### Fixed
- Fixed bug where an image that is rotated with EXIF data loses its rotation (#67)

## [2.0.5] - 2020-05-04

No changes

## [2.0.4] - 2020-05-04

No changes

## [2.0.3] - 2020-05-04

### Fixed
- Fixed bug with Serverless not excluding dev-dependencies and failing to deploy due to large file size
- Fixed bug with Sharp binaries (Thanks @bs-thomas) (#63)

## [2.0.3] - 2020-05-04

### Added
- Started new docs!
- Add `SLS_VALID_PATH_REGEX` setting to explicitly control what paths are allowed (Thanks @bs-thomas) (#52)
- Add support for deploying to custom region (#61)
- Added support for SSL & custom domain (Thanks @bs-thomas) (#47)
- Added `DEFAULT_CACHE_CONTROL` to allow settings a cache control value when one isn't set on the source object
- Added `settings.js` helper

### Fixed
- Error that could occur when cropping via focalpoint and a non-integer is encountered
- Errors would always return 500 instead of correct error code (#45)

## Changed
- Update platform from Node 10 to Node 12 (#59)
- Update Sharp from 0.23.4 to 0.25.2
- The proper response code is now returned on error
- checkHash now throws a 400 HashException when a security hash is not present
- Update Standard from 14.3.1 to 14.3.3
- Update Serverless from 1.60.5 to 1.69.0
- Update serverless-offline from 5.12.1 to 6.1.4
- Update Jest from 24.9.0 to 25.5.4
- Update aws-sdk from 2.600.0 to 2.668.0
- Update aws-sdk-mock from 4.5.0 to 5.1.0

## [2.0.2] - 2020-02-04

### Changed
- Improved speed of input buffering for `pngquant` by skipping compression

### Added
- `pngquant` speed can now be configured in settings, defaults to fastest speed

### Fixed
- `fill-color` for hex color codes now supports 3-,4-,6-, and 8-digit formats
- Incoming image size limitations can now be configured and enforced with environment variables, preventing large images
from timing out

## [2.0.1] - 2020-01-09

### Changed
- Sharp is now bundled as its own layer, removing the need for swapping between platforms for local development
- Updated to latest version of Sharp (0.23.4)

### Added
- Added support for `ar` param, allowing an aspect ratio (16:9) to be supplied

### Fixed
- Fixed a potential bug with automatic webp formatting in non-webp-supporting browsers

## [2.0.0] - 2019-12-09

### Changed
- A breaking change has been introduced in the way URLs are generated. Previously we were calling `encodeURIComponent`
on each part of the requested URL, we are now calling a modified `encodeURIComponent` that excludes some characters to
be compliant with Imgix's official php SDK:
```php
$path = preg_replace_callback("/([^\w\-\/\:@])/", function ($match) {
    return rawurlencode($match[0]);
}, $path);
```

See https://locutus.io/php/url/rawurlencode/ for details
 
### Added
- Added environment variable `SLS_IGNORE` with default value of `favicon.ico` 

## Fixed
- Issue with Imgix compatibility regarding URL encoding. See Changed section

## [1.0.2] - 2019-12-09
### Added
- Added support for `dpr`

### Changed
- Updated dependencies
- Added MIT license
- Values outside of explicit rule min and max will now normalize to that min or max

## [1.0.1] - 2019-11-21
### Fixed
- Fixed an error in RegEx hex color matching that could cause some colors to fail (#0000)

## [1.0.0] - 2019-10-25
### Release
- Initial stable release

## [1.0.0-beta.2] - 2019-08-31
### Changed
- Update Imgix schema 11.0.0 -> 11.1.1
- Security will now allow images without security hashes to passthrough if there are no edits

### Fixed
- Fixed issue where errors wouldn't properly pass-through to image

## [1.0.0-beta.1] - 2019-08-27
### Added
- Changelog

### Changed
- Dropped support for the `MAX_OUTPUT_HEIGHT` and `MAX_OUTPUT_WIDTH` to better adhere to the schema.json spec
- Remove unused dependency `serverless-plugin-simulate` to resolve security issue

### Fixed
- Ensure all file extensions are switched to lowercase before processing


src/helpers/settings.js
const SettingsException = require('../errors/SettingsException')

const TYPE_INTEGER = 'integer'
const TYPE_ARRAY_STRING = 'arraystring'
const TYPE_REGEX = 'regex'
const TYPE_STRING = 'string'

const settings = {
  DEFAULT_QUALITY: {
    default: 75,
    type: TYPE_INTEGER
  },
  DEFAULT_COMPRESS_QUALITY: {
    default: 75,
    type: TYPE_INTEGER
  },
  SLS_IGNORE: {
    default: '',
    type: TYPE_ARRAY_STRING
  },
  SLS_VALID_PATH_REGEX: {
    default: '.*',
    type: TYPE_REGEX
  },
  MAX_IMAGE_WIDTH: {
    default: 2000,
    type: TYPE_INTEGER
  },
  MAX_IMAGE_HEIGHT: {
    default: 1000,
    type: TYPE_INTEGER
  },
  PNGQUANT_SPEED: {
    default: 10,
    type: TYPE_INTEGER
  },
  DEFAULT_CACHE_CONTROL: {
    default: '',
    type: TYPE_STRING
  },
  SOURCE_BUCKET: {
    default: '',
    type: TYPE_STRING
  },
  SECURITY_KEY: {
    default: '',
    type: TYPE_STRING
  },
  CUSTOM_DOMAIN: {
    default: '',
    type: TYPE_STRING
  }
}

/**
 * Gets a setting from the config
 * @param key
 * @return {string|null}
 */
exports.getSetting = function (key) {
  if (!(key in settings)) {
    throw new SettingsException()
  }
  let value = null
  if (key in process.env) {
    value = process.env[key]
  } else {
    value = settings[key].default
  }

  return processValue(key, value)
}

const processValue = function (setting, value) {
  switch (settings[setting].type) {
    case TYPE_STRING:
      return processString(value)
    case TYPE_INTEGER:
      return processInteger(value)
    case TYPE_ARRAY_STRING:
      return processStringArray(value)
    case TYPE_REGEX:
      return processRegExValue(value)
    default:
      throw new SettingsException()
  }
}

const processString = function (value) {
  if (value === '' || value == null) {
    return null
  }
  return value.toString()
}

const processInteger = function (value) {
  return parseInt(value)
}

const processStringArray = function (value) {
  return value.split(',')
}

const processRegExValue = function (value) {
  return new RegExp(value)
}
